cfgov/unprocessed/js/organisms/FilterableListControls.js

```diff
@@ -18,157 +17,254 @@ var Notification = require( '../molecules/Notification' );
+   */
+  function _getLabelText( field, selector, isInGroup ) {
+    var labelText = '';
+    var labelDom;
+
+    if ( isInGroup && !selector ) {
+      labelDom = getClosestElement( field, 'fieldset' );
+      if ( labelDom ) labelDom = labelDom.querySelector( 'legend' );
     } else {
-      _showSuccess();
+      selector = selector ||
+                 'label[for="' + field.getAttribute( 'id' ) + '"]';
+      labelDom = _form.querySelector( selector );
     }
+
+    if ( labelDom ) labelText = labelDom.textContent.trim();
+
+    return labelText;
   }
 
   /**
-   * Show error notification.
-  */
-  function _showError() {
-    _notification.setTypeAndContent( _notification.ERROR, 'Error!' );
-    _notification.show();
+   * Set the notification type, msg, and visibility.
+   * @param {String} type The type of notification to display.
+   * @param {String} msg The message to display in the notification.
+   * @param {String} methodName The method to use to control visibility
+                                of the notification.
+   */
+  function _setNotification( type, msg, methodName ) {
+    methodName = methodName || 'show';
+    _notification.setTypeAndContent( type, msg );
+    _notification[methodName]();
   }
 
   /**
-   * Show success notification.
-  */
-  function _showSuccess() {
-    _notification.setTypeAndContent( _notification.SUCCESS, 'Success!' );
-    _notification.show();
+   * Determines if you should validate a field.
+   * @param {Field} field A form field.
+   * @param {String} type The type of field.
+   * @param {Boolean} isInGroup A boolean that determines if field in a group.
+   * @returns {Boolean} Value indicating whether to validate a field.
+   */
+  function shouldValidateField( field, type, isInGroup ) {
+    var isDisabled = field.getAttribute( 'disabled' ) !== null;
+    var isIgnoreType = _defaults.ignoreFieldTypes.indexOf( type ) > -1;
+    var shouldValidate = isDisabled === false && isIgnoreType === false;
+
+    if ( shouldValidate && ( isInGroup || false ) ) {
+      var name = field.getAttribute( 'data-group' ) ||
+                 field.getAttribute( 'name' );
+      var isRequired = field.getAttribute( 'data-required' ) !== null;
+      if ( _fieldGroups.indexOf( name ) > -1 || isRequired === false ) {
+        shouldValidate = false;
+      } else {
+        _fieldGroups.push( name );
+      }
+    }
+
+    return shouldValidate;
   }
 
   /**
    * Validate the fields of our form.
-   * @param  {Array} fields The list of input fields we're testing.
+   * @param {Array} fields A list of form fields.
    * @returns {Object}
    *   The tested list of fields broken into valid and invalid blocks.
-  */
+   */
   function _validateFields( fields ) {
-    var checkgroups = {};
     var validatedFields = {
-      valid:   [],
-      invalid: []
+      invalid: [],
+      valid:   []
     };
+    var validatedField;
+
+    _fieldGroups = [];
+
+    fields.forEach( function loopFields( field ) {
```

Comment:
Yeah, sorry to imply it's a performance tweak—this is more about disentangling the steps vs performance. If functions aren't nested it means they can be isolated. In `foreach` one function is nested inside the loop so there is a tight coupling between the steps (loop, do something on each item), whereas in a classic loop, it's a loose coupling. There's one step that calls another. One function is for looping over the array, and another is for setting the value. For testing, having the two functions separate means they can be more easily tested in isolation, easier to document (not that we're utilizing jsdocs yet), and easier to refactor in isolation (e.g. made public, shared between functions, etc.). But again as private methods it may not matter.